### PR TITLE
Add partion support

### PR DESCRIPTION
**Pull request by zetaab**
this adds partion support to first functions of go-bigip library. Please note that this is not backward compatible because new variables are added to function calls.

If partion is `""` it is treated like currenly known as Common

### CODE REVIEW

**Comment by kenmaglio** [comment]
So @zetaab you do know you can specify partitions already on all the calls that accept full path and the functions will already support them.  You simply use folder structure ( all they really are under the hood on the F5). 

so   /k8s/myapp_80_vs   
k8s partition
myapp_80_vs = object

We've fully utilized this package controlling all objects within a partition just fine thus far. 
Just remember to reference an object FullPath = /partition/name

The code will change it before the API call to FullPath  ~partition~name

**Comment by zetaab** [comment]
@kenmaglio yes, I noticed that some functions works like that. However, there are quite many functions which do not work. Example: https://github.com/ElisaOyj/openshift-lb-controller/pull/8

File: ltm.go [hunk 1]
```diff
@@ -1644,11 +1648,16 @@ func (b *BigIP) PoolMembers(name string) (*PoolMembers, error) {
...
 // of <node>:<port>, i.e.: "web-server1:443".
-func (b *BigIP) AddPoolMember(pool, member string) error {
+func (b *BigIP) AddPoolMember(pool, member string, partition string) error {
 	config := &poolMember{
 		Name: member,
 	}
-
+	if len(partition) > 0 {
+		// node partition
+		config.Partition = partition
```
Comment (zetaab) [inline ltm.go]: In this function, the node needs partition as POST argument. There is no way to do this without this.

File: ltm.go [hunk 2]
```diff
@@ -1738,11 +1753,14 @@ func (b *BigIP) PoolMemberStatus(pool string, member string, state string, owner
...
 
 // CreatePool adds a new pool to the BIG-IP system by name.
-func (b *BigIP) CreatePool(name string) error {
+func (b *BigIP) CreatePool(name string, partition string) error {
 	config := &Pool{
 		Name: name,
 	}
 
+	if len(partition) > 0 {
+		config.Partition = partition
```
Comment (zetaab) [inline ltm.go]: same here, no possibility to without partition specified in POST.

File: ltm.go [hunk 3]
```diff
@@ -1951,7 +1975,9 @@ func (b *BigIP) CreateMonitor(name, parent string, interval, timeout int, send,
 		SendString:    send,
 		ReceiveString: receive,
 	}
-
+	if len(partition) > 0 {
+		config.Partition = partition
```
Comment (zetaab) [inline ltm.go]: same here, no possibility to without partition specified in POST.

File: ltm.go [hunk 4]
```diff
@@ -1997,15 +2026,24 @@ func (b *BigIP) ModifyMonitor(name, monitorType string, config *Monitor) error {
...
 // AddMonitorToPool assigns the monitor, <monitor> to the given <pool>.
-func (b *BigIP) AddMonitorToPool(monitor, pool string) error {
+func (b *BigIP) AddMonitorToPool(monitor, pool string, partition string) error {
 	config := &Pool{
 		Monitor: monitor,
 	}
 
+	if len(partition) > 0 {
+		// node partition
+		config.Partition = partition
```
Comment (zetaab) [inline ltm.go]: same here, no possibility to without partition specified in POST.

**Comment by zetaab** [comment]
@kenmaglio check my comments above ^. I do agree that some things you can do without changing code this much. However, I would like to see functions working like similar way that is why I added partition to functions as well which does not need it.

File: ltm.go (hunk 1 above)
Comment (streamingmovieguy) [reply ltm.go]: So actually, we have fully implemented Pool and Pool Member creation in a partition without this modification.

What you need to look at is the function getForEntity 
This calls iControlPath on the dynamic path parameter
This function takes the dynamic path and breaks apart each param, and looks if any of those individual params includes "/"  and replaces with "~"  --- what you are doing to add partition support - it's already there my friend.

func (b *BigIP) iControlPath(parts []string) string {
	var buffer bytes.Buffer
	var lastPath int
	if strings.HasPrefix(parts[len(parts)-1], "?") {
		lastPath = len(parts) - 2
	} else {
		lastPath = len(parts) - 1
	}
	for i, p := range parts {
		buffer.WriteString(strings.Replace(p, "/", "~", -1))
		if i < lastPath {
			buffer.WriteString("/")
		}
	}
	return buffer.String()
}

Comment (kenmaglio) [reply ltm.go]: So actually, we have fully implemented Pool and Pool Member creation in a partition without this modification.

What you need to look at is the function getForEntity 
This calls iControlPath on the dynamic path parameter
This function takes the dynamic path and breaks apart each param, and looks if any of those individual params includes "/"  and replaces with "~"  --- what you are doing to add partition support - it's already there my friend.

func (b *BigIP) iControlPath(parts []string) string {
	var buffer bytes.Buffer
	var lastPath int
	if strings.HasPrefix(parts[len(parts)-1], "?") {
		lastPath = len(parts) - 2
	} else {
		lastPath = len(parts) - 1
	}
	for i, p := range parts {
		buffer.WriteString(strings.Replace(p, "/", "~", -1))
		if i < lastPath {
			buffer.WriteString("/")
		}
	}
	return buffer.String()
}

File: ltm.go (hunk 2 above)
Comment (kenmaglio) [reply ltm.go]: So if you need to Create a Pool to specify a partition -- look at  AddPool

```go
f5Pool = &bigip.Pool{
    Name:      f5PoolName,
    Partition: s.partition,
}
err = s.f5.AddPool(f5Pool)
```

File: ltm.go (hunk 3 above)
Comment (kenmaglio) [reply ltm.go]: Get and Delete Pool   100% support  "/partition/nodename"  we have implemented this with 100% success

Again you need to look at  AddMonitor   not Create if you want to pass full spec  including partition

File: ltm.go (hunk 4 above)
Comment (kenmaglio) [reply ltm.go]: 100% DeleteMonitor  and PatchMonitor will work with  "/partition/monitorname"  as they as well call iControlPath on the path

**Comment by kenmaglio** [comment]
My other worry on this - as you've updated the tests, is that this would potentially be a breaking change for others that use this already, requiring the additional argument to be passed. 

Not sure that's a great thing. 
https://play.golang.org/p/UfRqe1fUId3

If I had all my code using this example with only the one param, now with this change it'll break.

**Comment by kenmaglio** [comment]
Maybe something like this... but even this is kinda ... grrrrr

https://play.golang.org/p/g6ATs--oIhc

**Comment by zetaab** [comment]
@kenmaglio Thanks for the tips. I can try to modify our code to work without modifications to this lib.

**Comment by kenmaglio** [comment]
If you need and help let me know
 We worked through a lot recently. :)

**Comment by zetaab** [comment]
@kenmaglio I got it working! Big thanks! So this pr is not needed if someone is interested how it is done https://github.com/ElisaOyj/openshift-lb-controller/pull/11

**Comment by kenmaglio** [comment]
@zetaab Awesome man!!!

**Pull request closed without merging**